modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,628 @@
+use core::fmt;
+use core::ops::Add;
+use core::str::FromStr;
+
+use derive_more::{Display, From};
+use ibc_proto::cosmos::base::v1beta1::Coin as RawCoin;
+use ibc_proto::ibc::applications::transfer::v1::DenomTrace as RawDenomTrace;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use subtle_encoding::hex;
+
+use super::error::Error;
+use crate::bigint::U256;
+use crate::core::ics24_host::identifier::{ChannelId, PortId};
+use crate::prelude::*;
+use crate::serializers::serde_string;
+
+const IBC_DENOM_PREFIX: &str = "ibc";
+
+/// A `Coin` type with fully qualified `DenomTrace`.
+pub type PrefixedCoin = Coin<DenomTrace>;
+
+/// A `Coin` type with a `HashedDenom`.
+pub type HashedCoin = Coin<HashedDenom>;
+
+/// A `Coin` type with an unprefixed denomination.
+pub type BaseCoin = Coin<Denom>;
+
+/// Base denomination type
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize, Display)]
+#[serde(transparent)]
+pub struct Denom(String);
+
+impl FromStr for Denom {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.trim().is_empty() {
+            Err(Error::empty_base_denom())
+        } else {
+            Ok(Denom(s.to_owned()))
+        }
+    }
+}
+
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+pub struct TracePrefix {
+    port_id: PortId,
+    channel_id: ChannelId,
+}
+
+impl TracePrefix {
+    pub fn new(port_id: PortId, channel_id: ChannelId) -> Self {
+        Self {
+            port_id,
+            channel_id,
+        }
+    }
+}
+
+impl fmt::Display for TracePrefix {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}/{}", self.port_id, self.channel_id)
+    }
+}
+
+/// A full trace path modelled as a collection of `TracePrefix`s.
+#[derive(Clone, Debug, Default, Eq, PartialEq, PartialOrd, Ord, From)]
+pub struct TracePath(Vec<TracePrefix>);
+
+impl<'a> TryFrom<Vec<&'a str>> for TracePath {
+    type Error = Error;
+
+    fn try_from(v: Vec<&'a str>) -> Result<Self, Self::Error> {
+        if v.len() % 2 != 0 {
+            return Err(Error::invalid_trace_length(v.len()));
+        }
+
+        let mut trace = vec![];
+        let id_pairs = v.windows(2).map(|paths| (paths[0], paths[1]));
+        for (pos, (port_id, channel_id)) in id_pairs.step_by(2).enumerate() {
+            let port_id =
+                PortId::from_str(port_id).map_err(|e| Error::invalid_trace_port_id(pos, e))?;
```

Comment:
The current error message prints out the `TracePrefix` index. It would be better & sufficient to print out the failing port, perhaps with its channel too?

Same goes for when the channel id parse fails.

Reply:
I think the `FromStr` impls of both `PortId` and `ChannelId` already return errors that print the identifiers along with a reason for parse failure.

Reply:
Oh, and `flex-error` would stack all these errors so that what the user ultimately sees is both the port/channel id, as well as this index message? If so, then yep let's leave it as is.